synapse/http/site.py

```diff
@@ -357,7 +365,20 @@ def connectionLost(self, reason: Union[Failure, Exception]) -> None:
                     {"event": "client connection lost", "reason": str(reason.value)}
                 )
 
-            if not self._is_processing:
+            if self._is_processing:
+                if self.is_render_cancellable:
+                    if self.render_deferred is not None:
+                        # Throw a cancellation into the request processing, in the hope
+                        # that it will finish up sooner than it normally would.
+                        # The `self.processing()` context manager will call
+                        # `_finished_processing()` when done.
+                        self.render_deferred.cancel()
```

Comment:
It turns out `Deferred.cancel()` is a lot like `Deferred.callback()/errback()` in that it will trash the logging context:
it can resume a coroutine, which will restore its own logging context, then run:
* until it blocks, setting the sentinel context
* or until it terminates, setting the context it was started with

So we need to wrap it in `with PreserveLoggingContext():`, like we do with `.callback()`:
```suggestion
                        with PreserveLoggingContext():
                            self.render_deferred.cancel()
```